### PR TITLE
Llama Email Agent

### PR DESCRIPTION
**Pull request by jeffxtang**
# What does this PR do?

An app that shows how to build a Gmail agent app powered by Llama 3.1 8B running locally via Ollama, with custom tool calling natively supported in Llama 3.1. 

Fixes # (issue)

## Feature/Issue validation/testing

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
Logs for Test A

- [ ] Test B
Logs for Test B

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

**Review by heyjustinai — CHANGES_REQUESTED** [review]
i like this use case! can you also add a test plan in PR please (screen shot or video of the agent working)

File: recipes/use_cases/gmail_agent/README.md [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+This Gmagent app shows how to build a Gmail agent app powered by Llama 3.1 8B running locally via Ollama (for privacy concern since Gamgent is about your Gmail). We'll start with building from scratch a basic agent with custom tool calling natively supported in Llama 3.1. The end goal is to cover all components of a production-ready agent app, acting as an assistant to your Gmail, with great user experience: intuitive, engaging, efficient and reliable.
+
+Currently implemented features of Gmagent include:
+* search for emails and attachments
+* get email detail
+* reply to a specific email 
+* forward an email
+* get summary of a PDF attachment
+* draft and send an email
```
[the remark this answers is not in the document]
Comment (jeffxtang) [reply recipes/use_cases/gmail_agent/README.md:11]: The example command line inputs and outputs are in examples_log.txt

**Comment by jeffxtang** [comment]
> i like this use case! can you also add a test plan in PR please (screen shot or video of the agent working)
Thanks for the review. Adding a test plan would be a bit too much work at this stage, but I'll add this task in the TODO. Yes some screenshots and demo video would be nice - will add them.

File: recipes/use_cases/gmail_agent/gmagent.py [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+    
+
+def get_pdf_summary(file_name):
+    text = pdf2text(file_name)
+    print("Calling Llama to generate a summary...")
+    response = llama31(text, "Generate a summary of the input text in 5 sentences.")
+    return response
+
+    
+def get_email_attachments(message_id, mime_type='application/pdf'):
```
[the remark this answers is not in the document]
Comment (jeffxtang) [reply recipes/use_cases/gmail_agent/gmagent.py:425]: Currently the app can only check for the default PDF attachments so if the user asks "do i have any word doc attachments larger than 5mb", the tool calling parameter will be "has:attachment filename:type:word larger:5mb" - more prompt engineering and pre-processing will be needed to handle this. For now I'll add another item in the TODO list.

File: recipes/use_cases/gmail_agent/main.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+from gmagent import *
+from functions_prompt import system_prompt
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Set email address")
+    parser.add_argument("--user_email", type=str, required=True, help="Your Gmail address")
+    args = parser.parse_args()
+
+    gmagent.set_email_service(args.user_email)
```
[the remark this answers is not in the document]
Comment (jeffxtang) [reply recipes/use_cases/gmail_agent/main.py:12]: what parser_args already has error handling to make sure the email argument is provided - you mean to additionally check if the email entered is valid?

File: recipes/use_cases/gmail_agent/gmagent.py [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+        if header['name'] == 'Subject':
+            subject = header['value']
+        elif header['name'] == 'From':
+            sender = header['value']    
+
+    internal_date = message.get('internalDate')  
+    utc_time = datetime.fromtimestamp(int(internal_date) / 1000, tz=timezone.utc)
+    
+    # Convert UTC to the specified timezone
+    local_timezone = pytz.timezone("America/Los_Angeles")
```
[the remark this answers is not in the document]
Comment (jeffxtang) [reply recipes/use_cases/gmail_agent/gmagent.py:143]: good suggestion but the app is intended to show some real custom tool calling then the llama stack support and more advanced agent capabilities (multi-step reasoning and memory...) so i'll likely leave those few constants as is for now and will some come code refactoring in the future (or create such a class while adding more features).

File: recipes/use_cases/gmail_agent/gmagent.py [hunk 5]
```diff
@@ -0,0 +1,624 @@
...
+
+def num_of_emails(query=''):
+    response = service.users().messages().list(
+        userId='me', 
+        q=query, 
+        maxResults=1).execute()
+    return response.get('resultSizeEstimate', 0)
+
+
+def list_emails(query='', max_results=100):
```
[the remark this answers is not in the document]
Comment (jeffxtang) [reply recipes/use_cases/gmail_agent/gmagent.py:56]: save as above. will do it in the future

File: recipes/use_cases/gmail_agent/gmagent.py [hunk 6]
```diff
@@ -0,0 +1,624 @@
...
+    print("Calling Llama to generate a summary...")
+    response = llama31(text, "Generate a summary of the input text in 5 sentences.")
+    return response
+
+    
+def get_email_attachments(message_id, mime_type='application/pdf'):
+    attachments = []
+
+    # Helper function to process email parts
+    def process_parts(parts):
```
[the remark this answers is not in the document]
Comment (jeffxtang) [reply recipes/use_cases/gmail_agent/gmagent.py:429]: agreed - but for now i'm trying to focus on agent specific implementation (have an item for attachment type handling in TODOs).

File: recipes/use_cases/gmail_agent/gmagent.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeffxtang) [reply recipes/use_cases/gmail_agent/gmagent.py:425]: added some screenshots (with emails masked - won't add a video demo for privacy reason:)

File: recipes/use_cases/gmail_agent/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (heyjustinai) [reply recipes/use_cases/gmail_agent/main.py]: I was thinking about adding error handling  to handle scenarios when it fails to connecting to Gmail API or saving token files - feel free to ignore if not applicable

[the remark this answers is not in the document]
Comment (jeffxtang) [reply recipes/use_cases/gmail_agent/main.py]: Ah if an incorrect token is entered and it fails to saving, an error "Malformed auth code" will now appear before exiting the program. 

When it fails to connecting to Gmail API (the only case I see is the missing of credentials.json file, described in README - but if the Gmail API service is down then typical network connection errors would occur), an error "No such file or directory: 'credentials.json'" will appear now.

[the remark this answers is not in the document]
Comment (heyjustinai) [reply recipes/use_cases/gmail_agent/main.py]: awesome, thanks for including these changes

**Review by heyjustinai — APPROVED** [review]

**Pull request merged**